bin/comment_on_pr

```diff
@@ -1,14 +1,125 @@
 #!/bin/bash -eu
 
-PR_MESSAGE_FILE=$1
-GITHUB_TOKEN=$2
+# This script is used to comment on a pull request, and must be run in a buldkite PR step.
+#
+# Usage:
+#
+#    # Post a new comment to the PR
+#    comment_on_pr "This is a comment" <github-token>
+#
+#    # Post a new comment which can be uniqly identifiered by the given comment id,
+#    # or update the existing comment that's associated with the given comment id.
+#    # You can choose to delete the existing comment if it exists by passing `--if-exist delete`.
+#    comment_on_pr --id comment-id "This is a comment" [--if-exist update|delete] <github-token>
+#
+#    # Delete the comment associated with the given comment id
+#    comment_on_pr --id comment-id <github-token>
+#
+# Please note the comment argument can be a path to a file containing the comment, or a markdown.
 
-if ! command -v gh &> /dev/null; then
-	brew install gh
+# Check dependencies and print their versions for diagnosis purposes
+cat <<EOF
+> jq --version
+$(jq --version)
+
+> curl --version
+$(curl --version)
+
+EOF
+
+if [[ "${BUILDKITE_PULL_REQUEST:-invalid}" == "invalid" ]]; then
+    echo "Error: this tool can only be called from a Buildkite PR job"
+    exit 1
+fi
+
+# Default options
+opt_comment_id=""
+opt_if_exist="update"
+
+while [[ "$#" -gt 0 ]]; do
+    case $1 in
+        --id) opt_comment_id="$2"; shift ;;
+        --if-exist) opt_if_exist="$2"; shift ;;
+        --*) echo "Unknown option: $1"; exit 1 ;;
+        *) break ;;
+    esac
+    shift
+done
+
+case "$#" in
+    1)
+        arg_pr_comment=""
+        arg_github_token=$1
+        ;;
+    2)
+        if [[ -f "$1" ]]; then
+            arg_pr_comment=$(cat "$1")
+        else
+            arg_pr_comment="$1"
+        fi
+
+        arg_github_token=$2
+        ;;
+    *) echo "Error: invalid comment argument"; exit 1 ;;
+esac
+
+github_user=$(basename "$(dirname "${BUILDKITE_PULL_REQUEST_REPO%.git}")")
+github_repo=$(basename "${BUILDKITE_PULL_REQUEST_REPO%.git}")
+issues_endpoint="https://api.github.com/repos/${github_user}/${github_repo}/issues"
+
+# Find the existing PR comment
+if [[ -n "$opt_comment_id" ]]; then
+    # Do not change this variable value, it's used to identify the comments posted by this utility.
+    comment_body_id="<!-- DO NOT REMOVE ci-toolkit-comment-identifier: $opt_comment_id -->"
+    existing_comment_id="$(curl -s \
+        -H "Authorization: token ${arg_github_token}" \
+        -H "Accept: application/vnd.github+json" \
+        "$issues_endpoint/${BUILDKITE_PULL_REQUEST}/comments" \
+        | jq ".[] | select(.body | contains(\"$comment_body_id\")) | .id")"
```

Comment:
🛑  Confirmed the issue:

```
$ bin/comment_on_pr --id 'Olivier aka "AliSoftware" is testing' "@AliSoftware here Testing with a \"tricky\" comment ID full of spaces and quotes." $GITHUB_TOKEN
> jq --version
jq-1.6

> curl --version
curl 7.87.0 (x86_64-apple-darwin22.0) libcurl/7.87.0 (SecureTransport) LibreSSL/3.3.6 zlib/1.2.11 nghttp2/1.51.0
Release-Date: 2022-12-21
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS GSS-API HSTS HTTP2 HTTPS-proxy IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL threadsafe UnixSockets

jq: error: syntax error, unexpected IDENT, expecting ';' or ')' (Unix shell quoting issues?) at <top-level>, line 1:
.[] | select(.body | contains("<!-- DO NOT REMOVE ci-toolkit-comment-identifier: Olivier aka "AliSoftware" is testing -->")) | .id                                                                                              
jq: 1 compile error
```

Sure thing it might be quite unlikely that we'd use a tricky comment ID with quotes and all, at least when we're using explicit/hardcoded comment IDs. But I can see a case where the comment ID could be derived from something dynamic, like `terraform-plan-$PROJECT_NAME` but then `$PROJECT_NAME` might contain special characters like quotes, backslash, etc.

We had a similar case for when I implemented [`annotate_test_failures`](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/blob/trunk/bin/annotate_test_failures) where the `--context` parameter for the annotation to add/update was derived and built from the buildkite step name amongst other things, which could contain special characters as well; so it's not like it's _totally_ unlikely to happen…

---

I haven't tried but I think that using `jq --arg body "$comment_body_id"` could be a possible solution to solve this?

Reply:
This issue actually came to me, but I decided to ignore it. It's probably looking for trouble if the caller decides to use "unusual" identifier, like quotes as you mentioned, newline potentially, or with `-->` in it. Reason I decided to ignore it is because the comment id is a hardcoded text created by developers, rather than taken from a user input. I don't think we would go look for troubles by using special characters in the code.

However, I think it'd be a good idea to at least perform a simple validation, like only `[0-9a-zA-Z\-_]` is acceptable characters.

Reply:
⬆️ is implemented in https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/51/commits/0a74452d45f0e755669830bcc3995519109f5376

Reply:
> Reason I decided to ignore it is because the comment id is a hardcoded text created by developers, rather than taken from a user input

Ideally yes I agree. But in practice I can see cases like the ID being `report-test-failures-$BUILDKITE_STEP` then if the buildkite step name contains spaces it will cause issues…?

But maybe it's a non issue after all and we can keep it as a limitation/restriction for now and improve later if we really need such use case. Limiting to `[0-9a-zA-Z_-]` for now is probably a good start indeed.